### ISSUE TITLE
Eslint confuses prop type

### ISSUE THREAD

**Issue by donalatorre** [yannickcr/eslint-plugin-react#2783]
Consider this simple Component, made with React and flow. It renders a list, using a custom type `SimpleType`.

```
// @flow
import React from 'react'

type SimpleType = {
  attribute: number
};

const simpleArray: Array<SimpleType> = [{ attribute: 1 }, { attribute: 2 }, { attribute: 3 }]

function Component () {
  const mappingFunction = ({ attribute }: SimpleType, idx: number) => (
    <li>
      Index
      {' '}
      {idx}
      {' '}
      has an instance of simpletype with
      {' '}
      {attribute}
      {' '}
      as an attribute
    </li>
  )
  const listOfNumbers = simpleArray.map(mappingFunction)
  return (
    <ul>
      {listOfNumbers}
    </ul>
  )
}

export default Component

```

I'm getting the following eslint error:
`
5:3  error  'attribute' PropType is defined but prop is never used  react/no-unused-prop-types
`

But I'm not even using PropTypes. My guess is that it's confusing mappingFunction with a function component, and therefore confusing SimpleType with a PropType. But this shouldn't happen, mappingFunction can't be a component since it also receives multiple arguments. Hence, there's no reason to assume SimpleType is a PropType.

**Comment by ljharb** [comment]
All prop types are PropTypes, just less powerful :-)

mappingFunction is a function that takes a destructured object argument and returns jsx - is there a reason you're not declaring that inline?

Function components can receive two arguments, the second being `context`, which can be a number if `contextType` were provided.

Unfortunately component detection is very difficult. a PR to handle this case would be welcome.

**Comment by donalatorre** [comment]
In my original code, I wrote mappingFunction inline. I only wrote it separately in the issue description for clarity. But even when I place it inline, I get the eslint error.

Ok, the argument thing sounds fair. But there could be other ways to figure out it's not a component. Would it be too difficult to ignore inline functions that return JSX? Or ignoring components declared inside other components?
